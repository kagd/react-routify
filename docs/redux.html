
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <title>Redux Â· React Routify</title>
        <meta charset="UTF-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.0">
        
        
        
    
    <link rel="stylesheet" href="../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-prism/prism.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="patterns.html" />
    
    
    <link rel="prev" href="nesting.html" />
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        <li class="header">Table of Contents</li>
        
        
    
        <li class="chapter " data-level="1.1" data-path="../">
            
                <a href="../">
            
                    
                    Read Me
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="quick-start.html">
            
                <a href="quick-start.html">
            
                    
                    Quick Start
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="nesting.html">
            
                <a href="nesting.html">
            
                    
                    Nesting
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.4" data-path="redux.html">
            
                <a href="redux.html">
            
                    
                    Redux
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5" data-path="patterns.html">
            
                <a href="patterns.html">
            
                    
                    Route Patterns
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6" data-path="params.html">
            
                <a href="params.html">
            
                    
                    Route Parameters
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7" data-path="link.html">
            
                <a href="link.html">
            
                    
                    Link
            
                </a>
            

            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href=".." >Redux</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h1 id="redux">Redux</h1><p>React Routify supports Redux out of the box.</p><pre><code class="lang-js"><span class="token keyword">import</span> React <span class="token keyword">from</span> <span class="token string">&apos;react&apos;</span>
<span class="token keyword">import</span> ReactDOM <span class="token keyword">from</span> <span class="token string">&apos;react-dom&apos;</span>
<span class="token keyword">import</span> <span class="token punctuation">{</span> ReduxRouter<span class="token punctuation">,</span> routeReducer <span class="token punctuation">}</span> <span class="token keyword">from</span> <span class="token string">&apos;react-routify&apos;</span>
<span class="token keyword">import</span> <span class="token punctuation">{</span> createStore<span class="token punctuation">,</span> combineReducers <span class="token punctuation">}</span> <span class="token keyword">from</span> <span class="token string">&apos;redux&apos;</span>
<span class="token keyword">import</span> <span class="token punctuation">{</span> Provider <span class="token punctuation">}</span> <span class="token keyword">from</span> <span class="token string">&apos;react-redux&apos;</span>

<span class="token keyword">const</span> rootReducer <span class="token operator">=</span> <span class="token function">combineReducers</span><span class="token punctuation">(</span><span class="token punctuation">{</span>
  route<span class="token punctuation">:</span> routeReducer <span class="token comment" spellcheck="true">// property must be named `route`</span>
<span class="token punctuation">}</span><span class="token punctuation">)</span>

<span class="token keyword">const</span> store <span class="token operator">=</span> <span class="token function">createStore</span><span class="token punctuation">(</span>rootReducer<span class="token punctuation">)</span>

<span class="token keyword">const</span> Home <span class="token operator">=</span> <span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token operator">=</span><span class="token operator">&gt;</span> <span class="token operator">&lt;</span>div<span class="token operator">&gt;</span>Home<span class="token operator">&lt;</span><span class="token operator">/</span>div<span class="token operator">&gt;</span>
<span class="token keyword">const</span> About <span class="token operator">=</span> <span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token operator">=</span><span class="token operator">&gt;</span> <span class="token operator">&lt;</span>div<span class="token operator">&gt;</span>About<span class="token operator">&lt;</span><span class="token operator">/</span>div<span class="token operator">&gt;</span>

<span class="token keyword">const</span> routes <span class="token operator">=</span> <span class="token punctuation">[</span>
  <span class="token punctuation">{</span> path<span class="token punctuation">:</span> <span class="token string">&apos;/&apos;</span><span class="token punctuation">,</span> component<span class="token punctuation">:</span> Home <span class="token punctuation">}</span><span class="token punctuation">,</span>
  <span class="token punctuation">{</span> path<span class="token punctuation">:</span> <span class="token string">&apos;/about&apos;</span><span class="token punctuation">,</span> component<span class="token punctuation">:</span> About <span class="token punctuation">}</span>
<span class="token punctuation">]</span>

<span class="token keyword">const</span> rootComponent <span class="token operator">=</span> <span class="token punctuation">(</span>
  <span class="token operator">&lt;</span>Provider store<span class="token operator">=</span><span class="token punctuation">{</span>store<span class="token punctuation">}</span><span class="token operator">&gt;</span>
    <span class="token operator">&lt;</span>ReduxRouter routes<span class="token operator">=</span><span class="token punctuation">{</span>routes<span class="token punctuation">}</span> store<span class="token operator">=</span><span class="token punctuation">{</span>store<span class="token punctuation">}</span> <span class="token operator">/</span><span class="token operator">&gt;</span>
  <span class="token operator">&lt;</span><span class="token operator">/</span>Provider<span class="token operator">&gt;</span>
<span class="token punctuation">)</span>

ReactDOM<span class="token punctuation">.</span><span class="token function">render</span><span class="token punctuation">(</span>
  rootComponent<span class="token punctuation">,</span>
  document<span class="token punctuation">.</span><span class="token function">getElementById</span><span class="token punctuation">(</span><span class="token string">&apos;root&apos;</span><span class="token punctuation">)</span>
<span class="token punctuation">)</span></code></pre>
<p>Route information is now available on <code>store.route</code>.</p>
                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                <a href="nesting.html" class="navigation navigation-prev " aria-label="Previous page: Nesting">
                    <i class="fa fa-angle-left"></i>
                </a>
                
                
                <a href="patterns.html" class="navigation navigation-next " aria-label="Next page: Route Patterns">
                    <i class="fa fa-angle-right"></i>
                </a>
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"Redux","level":"1.4","depth":1,"next":{"title":"Route Patterns","level":"1.5","depth":1,"path":"docs/patterns.md","ref":"docs/patterns.md","articles":[]},"previous":{"title":"Nesting","level":"1.3","depth":1,"path":"docs/nesting.md","ref":"docs/nesting.md","articles":[]},"dir":"ltr"},"config":{"plugins":["edit-link","prism","-highlight","github","anchorjs"],"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"pluginsConfig":{"prism":{},"github":{"url":"https://github.com/rwhitmire/react-routify"},"search":{},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"fontsettings":{"theme":"white","family":"sans","size":2},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"edit-link":{"label":"Edit This Page","base":"https://github.com/rwhitmire/react-routify/tree/master"},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false},"anchorjs":{}},"theme":"default","pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"variables":{},"title":"React Routify","gitbook":"3.x.x"},"file":{"path":"docs/redux.md","mtime":"2016-09-26T12:19:02.164Z","type":"markdown"},"gitbook":{"version":"3.2.0","time":"2016-09-29T03:24:27.281Z"},"basePath":"..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../gitbook/gitbook.js"></script>
    <script src="../gitbook/theme.js"></script>
    
        
        <script src="../gitbook/gitbook-plugin-edit-link/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-github/plugin.js"></script>
        
    
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/3.1.1/anchor.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-anchorjs/anchor-style.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

