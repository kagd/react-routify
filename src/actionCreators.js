export const LOCATION_CHANGE = '@@router/LOCATION_CHANGE'
export function changeLocation(component, context) {
  return {
    type: LOCATION_CHANGE,
    component,
    context
  }
}
