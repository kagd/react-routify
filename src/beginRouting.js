import page from 'page'

const beginRouting = () => {
  page()
}

export default beginRouting
