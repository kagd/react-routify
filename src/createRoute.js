import page from 'page'

const createRoute = (path, handler) => {
  page(path, handler)
}

export default createRoute
