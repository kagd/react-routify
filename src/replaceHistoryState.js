const replaceHistoryState = path => {
  history.replaceState(history.state, null, path)
}

export default replaceHistoryState
