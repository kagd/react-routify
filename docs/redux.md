# Redux

React Routify supports Redux out of the box.

```js
import React from 'react'
import ReactDOM from 'react-dom'
import { ReduxRouter, routeReducer } from 'react-routify'
import { createStore, combineReducers } from 'redux'
import { Provider } from 'react-redux'

const rootReducer = combineReducers({
  route: routeReducer // property must be named `route`
})

const store = createStore(rootReducer)

const Home = () => <div>Home</div>
const About = () => <div>About</div>

const routes = [
  { path: '/', component: Home },
  { path: '/about', component: About }
]

const rootComponent = (
  <Provider store={store}>
    <ReduxRouter routes={routes} store={store} />
  </Provider>
)

ReactDOM.render(
  rootComponent,
  document.getElementById('root')
)
```

Route information is now available on `store.route`.
