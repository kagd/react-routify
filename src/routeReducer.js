import { LOCATION_CHANGE } from './actionCreators'

export default function(state = { path: '/', _private: {}}, action) {
  switch(action.type) {
  case LOCATION_CHANGE:
    return {
      _private: { component: action.component },
      ...action.context
    }
  default:
    return state
  }
}
