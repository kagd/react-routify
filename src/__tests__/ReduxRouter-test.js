jest.mock('../createRoute')
jest.mock('../beginRouting')
jest.mock('../replaceHistoryState')
jest.mock('../store')

import { createStore, combineReducers } from 'redux'
import React from 'react'
import renderer from 'react-test-renderer'
import ReduxRouter from '../ReduxRouter'
import createRoute from '../createRoute'
import beginRouting from '../beginRouting'
import replaceHistoryState from '../replaceHistoryState'
import routeReducer from '../routeReducer'
import internalStore from '../store'
import { changeLocation } from '../actionCreators'

const createTestStore = () => {
  return  createStore(combineReducers({
    route: routeReducer
  }))
}

test('default state should render null', () => {
  const Home = () => <div>home</div>
  const routes = [{ path: '/', component: Home }]

  const component = renderer.create(
    <ReduxRouter routes={routes} store={createTestStore()} />
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('beginRouting should be called', () => {
  const Home = () => <div>home</div>
  const routes = [{ path: '/', component: Home }]
  renderer.create(<ReduxRouter routes={routes} store={createTestStore()} />)
  expect(beginRouting).toBeCalled()
})

test('navigate should update the component', () => {
  const createRouteCallbacks = {}

  createRoute.mockImplementation((path, callback) => {
    createRouteCallbacks[path] = callback
  })

  const Home = () => <div>home</div>
  const About = () => <div>about</div>

  const routes = [
    { path: '/', component: Home },
    { path: '/about', component: About }
  ]

  const component = renderer.create(
    <ReduxRouter routes={routes} store={createTestStore()} />
  )

  let tree

  createRouteCallbacks['/']({path: '/'})
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  createRouteCallbacks['/about']({path: '/about'})
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('route information should exist on store', () => {
  const createRouteCallbacks = {}

  createRoute.mockImplementation((path, callback) => {
    createRouteCallbacks[path] = callback
  })

  const Home = () => <div>Home</div>
  const routes = [{ path: '/', component: Home }]
  const store = createTestStore()

  renderer.create(
    <ReduxRouter routes={routes} store={store} />
  )

  createRouteCallbacks['/']({path: '/about'})
  expect(store.getState().route.path).toBe('/about')
})

test('state updates should trigger replaceHistoryState', () => {
  const Home = () => <div>Home</div>
  const routes = [{ path: '/', component: Home }]
  const store = createTestStore()

  renderer.create(
    <ReduxRouter routes={routes} store={store} />
  )

  store.dispatch(changeLocation(Home, {}))
  expect(replaceHistoryState).toBeCalled()
})

test('state updates should trigger store emitChange', () => {
  const Home = () => <div>Home</div>
  const routes = [{ path: '/', component: Home }]
  const store = createTestStore()

  renderer.create(
    <ReduxRouter routes={routes} store={store} />
  )

  store.dispatch(changeLocation(Home, {}))
  expect(internalStore.emitChange).toBeCalled()
})
