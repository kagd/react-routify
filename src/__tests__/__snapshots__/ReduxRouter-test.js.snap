exports[`test default state should render null 1`] = `null`;

exports[`test navigate should update the component 1`] = `
<div>
  home
</div>
`;

exports[`test navigate should update the component 2`] = `
<div>
  about
</div>
`;
